app/src/main/java/org/openobservatory/ooniprobe/domain/QueryDataSource.java

```diff
@@ -0,0 +1,73 @@
+package org.openobservatory.ooniprobe.domain;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+import androidx.paging.PagingSource;
+import androidx.paging.PagingState;
+
+import com.google.common.collect.Iterables;
+import com.raizlabs.android.dbflow.sql.language.SQLOperator;
+import com.raizlabs.android.dbflow.sql.language.SQLite;
+
+import org.openobservatory.ooniprobe.common.ThirdPartyServices;
+import org.openobservatory.ooniprobe.model.database.Result;
+import org.openobservatory.ooniprobe.model.database.Result_Table;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+
+import kotlin.coroutines.Continuation;
+
+public class QueryDataSource extends PagingSource<Integer, Result> {
+    private final String testGroupNameFilter;
+
+    public QueryDataSource(String testGroupNameFilter) {
+        this.testGroupNameFilter = testGroupNameFilter;
+    }
+
+    @Nullable
+    @Override
+    public Integer getRefreshKey(@NonNull PagingState<Integer, Result> pagingState) {
+        return 1;
+    }
+
+    @Nullable
+    @Override
+    public LoadResult.Page<Integer, Result> load(@NonNull LoadParams<Integer> loadParams, @NonNull Continuation<? super LoadResult<Integer, Result>> continuation) {
+        // Key may be null during a refresh, if no explicit key is passed into Pager
+        // construction. Use 0 as default, because our API is indexed started at index 0
+        int highestId = loadParams.getKey() != null ? loadParams.getKey() : 0;
+
+        List<SQLOperator> conditions = (testGroupNameFilter != null && !testGroupNameFilter.isEmpty())
+                ? Collections.singletonList(Result_Table.test_group_name.is(testGroupNameFilter))
+                : new ArrayList<>();
+        if (highestId != 0) {
+            conditions.add(Result_Table.id.lessThan(highestId));
+        }
+
+        List<Result> response = SQLite.select().from(Result.class)
+                .where(conditions.toArray(new SQLOperator[0]))
+                .limit(20)
+                .orderBy(Result_Table.id, false)
+                .queryList();
+
+        Integer prevKey;
+        try {
+            prevKey = (!response.isEmpty()) ? Objects.requireNonNull(Iterables.getFirst(response, null)).id - 20 : null;
+        } catch (Exception e) {
+            prevKey = null;
+            ThirdPartyServices.logException(e);
+        }
+        // This API defines that it's out of data when a page returns empty. When out of
+        // data, we return `null` to signify no more pages should be loaded
+        Integer nextKey = (!response.isEmpty()) ? Iterables.getLast(response).id : null;
```

Comment:
```suggestion
        Integer nextKey = (!response.isEmpty()) ? Iterables.getLast(response).id + 1 : null;
```

Conceptually, the next page seems to start at index `$last + 1` rather than at index `$last`. That said, I can see how smooth scrolling means that probably we want to keep the last element in scope.

Reply:
The operand used for the query is `lessThan` so using `Iterables.getLast(response).id + 1`  will mean we lose a record.